tools/c7n_azure/c7n_azure/functionapp_utils.py

```diff
@@ -33,6 +35,102 @@ def generate_machine_decryption_key():
         # randomly generated decryption key for Functions key
         return str(hexlify(os.urandom(32)).decode()).upper()
 
+    class FunctionAppInfrastructureParameters:
+        def __init__(self, group_name, location, app_insights_location, storage_name,
+                     service_plan_name, sku_name, sku_tier, webapp_name):
+            self.group_name = group_name
+            self.location = location
+            self.app_insights_location = app_insights_location
+            self.storage_name = storage_name
+            self.service_plan_name = service_plan_name
+            self.sku_name = sku_name
+            self.sku_tier = sku_tier
+            self.webapp_name = webapp_name
+
+    def deploy_infrastructure(self, parameters):
```

Comment:
I think this function would be a lot cleaner and easier to read if we pulled out each resource:
deploy_resource_group
deploy_storage
deploy_app_insights
deploy_service_plan 

One of the benefits of code over ARM templates :D